### PR TITLE
Accelerate Float16 and Bfloat16 GEMM functions

### PR DESCRIPTION
**Pull request by mikekgfb**
Accelerate GEMM for Half and BFloat by using Float GEMM functions

# ARM Measurements (MacOS Apple Silicon, M1 Max)
consistently advantageous

*before this PR*
notrans torch.float32 6.04 usec
notrans torch.float16 64.63 usec
notrans torch.bfloat16 63.33 usec
trans_a torch.float32 5.77 usec
trans_a torch.float16 69.56 usec
trans_a torch.bfloat16 317.89 usec
trans_b torch.float32 5.88 usec
trans_b torch.float16 48.79 usec
trans_b torch.bfloat16 51.88 usec

*after this PR*
notrans torch.float32 5.95 usec
notrans torch.float16 22.77 usec
notrans torch.bfloat16 22.70 usec
trans_a torch.float32 6.07 usec
trans_a torch.float16 22.13 usec
trans_a torch.bfloat16 22.32 usec
trans_b torch.float32 5.96 usec
trans_b torch.float16 22.81 usec
trans_b torch.bfloat16 22.92 usec

# x86 Measurements (AVX2)
advantageous for fp16 notrans/transa, degradation for transb
about same as before for bfloat except notrans case

*before*
notrans torch.float32 14.38 usec
notrans torch.float16 214.61 usec
notrans torch.bfloat16 36.16 usec
trans_a torch.float32 16.00 usec
trans_a torch.float16 328.54 usec
trans_a torch.bfloat16 38.14 usec
trans_b torch.float32 26.78 usec
trans_b torch.float16 78.91 usec
trans_b torch.bfloat16 37.55 usec

*after PR*
notrans torch.float32 14.98 usec
notrans torch.float16 97.27 usec
notrans torch.bfloat16 48.12 usec
trans_a torch.float32 14.84 usec
trans_a torch.float16 94.93 usec
trans_a torch.bfloat16 38.78 usec
trans_b torch.float32 26.38 usec
trans_b torch.float16 106.74 usec
trans_b torch.bfloat16 37.67 usec

Questionable whether we should do this on mobile for iOS because of the implied increase in memory foot print

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/CMakeLists.txt [hunk 1]
```diff
@@ -45,6 +45,11 @@ if(USE_ROCM)
   configure_file(hip/HIPConfig.h.in "${CMAKE_CURRENT_SOURCE_DIR}/hip/HIPConfig.h")
 endif()
 
+if(NOT MSVC)
+  # cudaProfilerInitialize must go away
+  set_source_files_properties(${CMAKE_CURRENT_SOURCE_DIR}/native/cpu/BlasKernel.cpp PROPERTIES COMPILE_FLAGS "-O3 -fvectorize -ftree-vectorize -ffp-contract=on -faggressive-loop-optimizations -march=native")
+endif()
+
```
Comment (malfet) [inline aten/src/ATen/CMakeLists.txt:52]: Those changes aren't really doing anything, are they?
```suggestion
```

File: aten/src/ATen/native/CPUBlas.cpp [hunk 2]
```diff
@@ -324,6 +324,46 @@ void gemm(
...
+
+#if AT_BUILD_WITH_BLAS()
+  if (use_blas_gemm(transa, transb, m, n, k, lda, ldb, ldc)) {
+    int m_ = m, n_ = n, k_ = k, lda_ = lda, ldb_ = ldb, ldc_ = ldc;
+    float alpha_ = alpha, beta_ = beta;
+    int a_size = (transa != TransposeType::NoTranspose?m_:k_) * lda;
+    std::vector<float> float_a(a, a + a_size);
+    int b_size = (transb != TransposeType::NoTranspose?k_:n_) * ldb;
+    std::vector<float> float_b(b, b + b_size);
+    int c_size = n_ * ldc;
```
Comment (malfet) [inline aten/src/ATen/native/CPUBlas.cpp:336]: Matrix size does not change, when it's transposed, does it?
```suggestion
    int a_size =  lda * k;
    std::vector<float> float_a(a, a + a_size);
    int b_size = ldb * n;
    std::vector<float> float_b(b, b + b_size);
    int c_size =  ldc * n;
```

File: aten/src/ATen/native/CPUBlas.cpp [hunk 3]
```diff
@@ -343,6 +383,46 @@ void gemm(
...
```
Comment (malfet) [inline aten/src/ATen/native/CPUBlas.cpp:395]:
```suggestion
    int a_size = k * lda;
    std::vector<float> float_a(a, a + a_size);
    int b_size = n * ldb;
    std::vector<float> float_b(b, b + b_size);
    int c_size = n * ldc;
```

File: aten/src/ATen/native/CPUBlas.cpp [hunk 4]
```diff
@@ -415,6 +517,28 @@ void gemm(
...
   }
 #endif
+
+#if AT_BUILD_WITH_BLAS()
+  if (use_blas_gemm(transa, transb, m, n, k, lda, ldb, ldc)) {
+    int m_ = m, n_ = n, k_ = k, lda_ = lda, ldb_ = ldb, ldc_ = ldc;
+    float alpha_ = alpha, beta_ = beta;
+    int a_size = (transa != TransposeType::NoTranspose?m_:k_) * lda;
+    std::vector<float> float_a(a, a + a_size);
+    int b_size = (transb != TransposeType::NoTranspose?k_:n_) * ldb;
```
Comment (swolchok) [inline aten/src/ATen/native/CPUBlas.cpp:527]: looks like @malfet changed the size computation in other places but not here; which is correct? probably explains our segfault, right?

Comment (swolchok) [reply aten/src/ATen/native/CPUBlas.cpp:527]: I think the right think might just be a_size = m * k, b_size = k * n, c_size = m * n? no messing around with lda/ldb/ldc

**Comment by swolchok** [comment]
I have a fixed version of this PR locally; I just made sure to compute a_size as m * k, b_size as k *n , and c_size as m * n everywhere, without regard to transposition.

It looks like this technique pays off on smaller matrix-vector sizes (e.g., the ones in the llm_experiments benchmark) but not larger ones, except for the case the benchmark labels `transa` (which I believe is actually `transb` from GEMM API perspective thanks to GEMM API assuming column order). Based on an accidental matrix-matrix experiment I ran, it might pay off for larger matrix-matrix multiplications, but we would need to pick a hardware target(s) to tune the threshold so I didn't run detailed experiments.

stock llm_experiments benchmark before (including #127033):

```
mv_nt    torch.float32    5.52 usec
mv_nt    torch.float16    8.93 usec
mv_nt   torch.bfloat16  194.83 usec
mv_ta    torch.float32    5.76 usec
mv_ta    torch.float16   25.06 usec
mv_ta   torch.bfloat16  103.72 usec
notrans  torch.float32    5.72 usec
notrans  torch.float16   26.54 usec
notrans torch.bfloat16   66.46 usec
trans_a  torch.float32    5.72 usec
trans_a  torch.float16   72.96 usec
trans_a torch.bfloat16  326.37 usec
trans_b  torch.float32    5.74 usec
trans_b  torch.float16    9.08 usec
trans_b torch.bfloat16   30.58 usec


Using FP16 accumulation
mv_nt    torch.float32    5.59 usec
mv_nt    torch.float16    8.92 usec
mv_nt   torch.bfloat16  195.12 usec
mv_ta    torch.float32    6.03 usec
mv_ta    torch.float16   21.64 usec
mv_ta   torch.bfloat16  109.69 usec
notrans  torch.float32    5.61 usec
notrans  torch.float16   19.98 usec
notrans torch.bfloat16   66.71 usec
trans_a  torch.float32    5.67 usec
trans_a  torch.float16   71.17 usec
trans_a torch.bfloat16  322.14 usec
trans_b  torch.float32    5.66 usec
trans_b  torch.float16    9.18 usec
trans_b torch.bfloat16   30.12 usec
```

with this PR:
```
mv_nt    torch.float32    5.50 usec
mv_nt    torch.float16   13.27 usec
mv_nt   torch.bfloat16  195.12 usec
mv_ta    torch.float32    5.76 usec
mv_ta    torch.float16   29.95 usec
mv_ta   torch.bfloat16  103.79 usec
notrans  torch.float32    5.66 usec
notrans  torch.float16   21.51 usec
notrans torch.bfloat16   22.04 usec
trans_a  torch.float32    5.82 usec
trans_a  torch.float16   22.21 usec
trans_a torch.bfloat16   22.35 usec
trans_b  torch.float32    5.75 usec
trans_b  torch.float16   22.14 usec
trans_b torch.bfloat16   22.94 usec


Using FP16 accumulation
mv_nt    torch.float32    5.55 usec
mv_nt    torch.float16    8.95 usec
mv_nt   torch.bfloat16  272.76 usec
mv_ta    torch.float32    5.76 usec
mv_ta    torch.float16   18.11 usec
mv_ta   torch.bfloat16  103.41 usec
notrans  torch.float32    5.61 usec
notrans  torch.float16   21.17 usec
notrans torch.bfloat16   21.14 usec
trans_a  torch.float32    5.63 usec
trans_a  torch.float16   20.99 usec
trans_a torch.bfloat16   21.31 usec
trans_b  torch.float32    5.84 usec
trans_b  torch.float16   23.26 usec
trans_b torch.bfloat16   22.51 usec
```

Summarizing result for "stock" suite:
notrans float16 improved slightly, bfloat16 improved significantly
trans_a float16 and bfloat16 improved significantly
trans_b float16 regressed, bfloat16 improved somewhat


I have a local patch to llm_experiments to use larger sizes.

Results from local patch before:
```
m=1024, n=1024, k=1
====================
notrans  torch.float32   25.62 usec
notrans  torch.float16  263.44 usec
notrans torch.bfloat16  250.36 usec
trans_a  torch.float32   28.12 usec
trans_a  torch.float16  699.90 usec
trans_a torch.bfloat16 1708.19 usec
trans_b  torch.float32   26.05 usec
trans_b  torch.float16   42.49 usec
trans_b torch.bfloat16  208.91 usec


Using FP16 accumulation
notrans  torch.float32   25.89 usec
notrans  torch.float16  259.57 usec
notrans torch.bfloat16  245.29 usec
trans_a  torch.float32   26.71 usec
trans_a  torch.float16  695.67 usec
trans_a torch.bfloat16 1702.69 usec
trans_b  torch.float32   26.29 usec
trans_b  torch.float16   43.05 usec
trans_b torch.bfloat16  209.37 usec
m=4100, n=4100, k=1
====================
notrans  torch.float32 2509.32 usec
notrans  torch.float16 4118.85 usec
notrans torch.bfloat16 4070.41 usec
trans_a  torch.float32 2699.20 usec
trans_a  torch.float16 17179.83 usec
trans_a torch.bfloat16 27366.06 usec
trans_b  torch.float32 2491.14 usec
trans_b  torch.float16  716.19 usec
trans_b torch.bfloat16 3902.81 usec


Using FP16 accumulation
notrans  torch.float32 2474.83 usec
notrans  torch.float16 4174.59 usec
notrans torch.bfloat16 4106.49 usec
trans_a  torch.float32 2839.11 usec
trans_a  torch.float16 17618.62 usec
trans_a torch.bfloat16 27850.44 usec
trans_b  torch.float32 2481.68 usec
trans_b  torch.float16  685.28 usec
trans_b torch.bfloat16 3873.80 usec
```

local patch suite after:
```
m=1024, n=1024, k=1
====================
notrans  torch.float32   27.26 usec
notrans  torch.float16  408.48 usec
notrans torch.bfloat16  407.29 usec
trans_a  torch.float32   28.11 usec
trans_a  torch.float16  399.67 usec
trans_a torch.bfloat16  462.84 usec
trans_b  torch.float32   27.36 usec
trans_b  torch.float16  403.54 usec
trans_b torch.bfloat16  405.55 usec


Using FP16 accumulation
notrans  torch.float32   26.92 usec
notrans  torch.float16  408.37 usec
notrans torch.bfloat16  433.04 usec
trans_a  torch.float32   29.30 usec
trans_a  torch.float16  409.88 usec
trans_a torch.bfloat16  400.41 usec
trans_b  torch.float32   27.24 usec
trans_b  torch.float16  400.09 usec
trans_b torch.bfloat16  494.35 usec
m=4100, n=4100, k=1
====================
notrans  torch.float32 2676.54 usec
notrans  torch.float16 9467.08 usec
notrans torch.bfloat16 9756.04 usec
trans_a  torch.float32 2763.41 usec
trans_a  torch.float16 10084.93 usec
trans_a torch.bfloat16 9845.06 usec
trans_b  torch.float32 2818.93 usec
trans_b  torch.float16 9805.75 usec
trans_b torch.bfloat16 9670.72 usec


Using FP16 accumulation
notrans  torch.float32 2730.35 usec
notrans  torch.float16 9861.93 usec
notrans torch.bfloat16 9872.48 usec
trans_a  torch.float32 3015.32 usec
trans_a  torch.float16 9956.29 usec
trans_a torch.bfloat16 9850.10 usec
trans_b  torch.float32 2708.49 usec
trans_b  torch.float16 9852.88 usec
trans_b torch.bfloat16 9825.65 usec
```

Summarizing result from local patch suite: trans_a improved for both 1024x1024x1 and 4100x4100x1, all other cases  regressed. Our "trans_a" kernel is clearly very bad; comparing to the trans_b kernel, I see that it is not multithreaded, so I will follow up to 1) fix that 2) try parallelizing the data type conversion in this PR.

I profiled the local suite in Instruments, and we are spending more time converting to float32 than we spend in BLAS. The conversion loop is nice and vectorized: for float16 it does two `ldp`, four pairs of `fcvtl`/`fcvtl2`, and four `stp`, and for bfloat16 is does two `ldp`, four `shll`, and two `stp`.

**Pull request closed without merging**